management/src/main/java/org/ehcache/management/registry/DefaultCapabilityBasedQuery.java

```diff
@@ -0,0 +1,302 @@
+
+/**
+ * @author Mathieu Carbou
+ */
+final class DefaultCapabilityBasedQuery implements CapabilityBasedQuery {
+
+  private final String capabilityName;
+  private final CapabilityManagement capabilityManagement;
+
+  DefaultCapabilityBasedQuery(CapabilityManagement capabilityManagement, String capabilityName) {
+    this.capabilityManagement = capabilityManagement;
+    this.capabilityName = capabilityName;
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistic(String statisticName) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, Collections.singletonList(statisticName));
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistics(String... statisticNames) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, Arrays.asList(statisticNames));
+  }
+
+  @Override
+  public StatisticQuery.Builder queryStatistics(Collection<String> statisticNames) {
+    return new StatisticQueryBuilder(capabilityManagement, capabilityName, statisticNames);
+  }
+
+  @Override
+  public CallQuery.Builder call(String methodName, String[] argClassNames, Object[] args) {
+    return new CallQueryBuilder(capabilityManagement, capabilityName, methodName, argClassNames, args);
+  }
+
+  @Override
+  public CallQuery.Builder call(String methodName) {
+    return call(methodName, new String[0], new Object[0]);
+  }
+
+  static class StatisticQueryBuilder implements StatisticQuery, StatisticQuery.Builder {
+
+    private final CapabilityManagement capabilityManagement;
+    private final String capabilityName;
+    private final Collection<String> statisticNames;
+    private final List<Map<String, String>> contexts;
+    private final long since;
+
+    StatisticQueryBuilder(CapabilityManagement capabilityManagement, String capabilityName, Collection<String> statisticNames) {
+      this(capabilityManagement, capabilityName, statisticNames, Collections.<Map<String, String>>emptyList(), 0);
+    }
+
+    private StatisticQueryBuilder(CapabilityManagement capabilityManagement, String capabilityName, Collection<String> statisticNames, List<Map<String, String>> contexts, long since) {
+      this.capabilityManagement = capabilityManagement;
+      this.capabilityName = capabilityName;
+      this.statisticNames = Collections.unmodifiableCollection(new LinkedHashSet<String>(statisticNames));
+      this.contexts = Collections.unmodifiableList(contexts);
+      this.since = since;
+    }
+
+    @Override
+    public StatisticQuery build() {
+      return this;
+    }
+
+    @Override
+    public StatisticQuery.Builder on(Map<String, String> context) {
+      if (!contexts.contains(context)) {
+        List<Map<String, String>> contexts = new ArrayList<Map<String, String>>(this.contexts);
+        contexts.add(Collections.unmodifiableMap(context));
+        return new StatisticQueryBuilder(capabilityManagement, capabilityName, statisticNames, contexts, since);
+      }
+      return this;
```

Comment:
This doesn't follow the Ehcache builder pattern which states that builders are immutable.